experiment/generate_tests.py

```diff
@@ -0,0 +1,233 @@
+    return substitute(job_name, field.get('args', []))
+
+
+def get_project_id(job_name):
+    """Returns the project id generated from the given job_name."""
+
+    return 'k8s-test-%s' % get_sha1_hash(job_name)[:10]
+
+
+def get_job_def(env_filename, args, sig_owners):
+    """Returns the job definition given the env_filename and the args."""
+
+    result = dict()
+    result['scenario'] = 'kubernetes_e2e'
+    result['args'] = []
+    result['args'].append('--env-file=%s' % env_filename)
+    result['args'].extend(args)
+    result['sigOwners'] = ['UNKNOWN'] if sig_owners is None else sig_owners
+    # Indicates that this job definition is auto-generated.
+    result['tags'] = ['generated']
+    result['_comment'] = 'AUTO-GENERATED - DO NOT EDIT.'
+    return result
+
+
+def write_env_file(output_dir, job_name, envs):
+    """Writes envs into a file in output_dir, and returns the file name."""
+
+    output_file = os.path.join(output_dir, job_name + '.env')
+    with open(output_file, 'w') as fp:
+        for env in envs:
+            fp.write(env + '\n')
+    return output_file
+
+
+def write_job_defs_file(output_dir, job_defs):
+    """Writes job definitions into a file in output_dir."""
+
+    output_file = os.path.join(output_dir, 'config.json')
+    with open(output_file, 'w') as fp:
+        json.dump(job_defs, fp, sort_keys=True, indent=4,
+                  separators=(',', ': '))
+        fp.write('\n')
+
+
+def generate_envs(job_name, common, cloud_provider, image, k8s_version,
+                  test_suite, job):
+    """Returns a list of envs fetched from the given fields."""
+
+    envs = ['# AUTO-GENERATED - DO NOT EDIT.']
+    envs.extend(get_envs('common', common))
+    envs.extend(get_envs('cloud provider', cloud_provider))
+    envs.extend(get_envs('image', image))
+    envs.extend(get_envs('k8s version', k8s_version))
+    envs.extend(get_envs('test suite', test_suite))
+    envs.extend(get_envs('job', job))
+    for env in envs:
+        if env.strip().startswith('PROJECT='):
+            break
+    else:
+        envs.extend(['', 'PROJECT=%s' % get_project_id(job_name)])
+    return envs
+
+
+def generate_args(job_name, common, cloud_provider, image, k8s_version,
+                  test_suite, job):
+    """Returns a list of args fetched from the given fields."""
+
+    args = []
+    args.extend(get_args(job_name, common))
+    args.extend(get_args(job_name, cloud_provider))
+    args.extend(get_args(job_name, image))
+    args.extend(get_args(job_name, k8s_version))
+    args.extend(get_args(job_name, test_suite))
+    args.extend(get_args(job_name, job))
+    return args
+
+
+def for_each_job(job_name, common, cloud_providers, images, k8s_versions,
+                 test_suites, jobs):
+    '''Returns the list of envs and args for each the job with the given
+    job_name.'''
+
+    fields = job_name.split('-')
+    if len(fields) != 7:
+        raise ValueError('Expected 7 fields in job name', job_name)
```

Comment:
The assumption is that this tool currently only supports jobs like `ci-kubernetes-e2e-<cloud_provider>-<image>-<k8s_version>-<test_suite>`. I plan to create more PRs to enhance it to support more job types later, such as node e2e and conformance tests. We can then document and validate the name. What we have in this PR is just a prototype.